// dockest/dist/test-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const rxjs_1 = require("rxjs");
const Logger_1 = require("./Logger");
exports.createRunner = (overrides) => ({
    commands: [],
    containerId: '',
    dependents: [],
    dockerComposeFileService: { image: 'node:10-alpine', ports: [{ published: 3000, target: 3000 }] },
    dockerEventStream$: new rxjs_1.ReplaySubject(),
    logger: new Logger_1.Logger('node'),
    readinessCheck: () => Promise.resolve(),
    serviceName: 'node',
    ...(overrides || {}),
});
exports.DOCKEST_SERVICE = {
    serviceName: 'redis',
};
exports.DOCKER_COMPOSE_FILE = {
    version: '3.8',
    services: {
        [exports.DOCKEST_SERVICE.serviceName]: {
            image: 'redis:5.0.3-alpine',
            ports: [
                {
                    published: 6379,
                    target: 6379,
                },
            ],
        },
    },
};
exports.MERGED_COMPOSE_FILES = `
services:
  redis:
    image: redis:5.0.3-alpine
    ports:
    - published: 6379
      target: 6379
version: '3.8'
`;
