// dockest/dist/utils/getOpts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const getRunMode_1 = require("./getRunMode");
const constants_1 = require("../constants");
exports.getOpts = (opts = {}) => {
    const { composeFile = 'docker-compose.yml', composeOpts: { alwaysRecreateDeps = false, build = false, forceRecreate = false, noBuild = false, noColor = false, noDeps = false, noRecreate = false, quietPull = false, } = {}, debug = false || process.argv.includes('dev') || process.argv.includes('debug'), dumpErrors = false, exitHandler = async () => Promise.resolve(), jestLib = require('jest'), jestOpts, jestOpts: { projects = ['.'], runInBand: runInBandJest = true } = {}, logLevel = constants_1.LOG_LEVEL.INFO, runInBand = true, } = opts;
    return {
        composeFile,
        composeOpts: {
            alwaysRecreateDeps,
            build,
            forceRecreate,
            noBuild,
            noColor,
            noDeps,
            noRecreate,
            quietPull,
        },
        debug,
        dumpErrors,
        exitHandler,
        mutables: {
            jestRanWithResult: false,
            runners: {},
            dockerEventEmitter: new events_1.EventEmitter(),
        },
        hostname: process.env.HOSTNAME || constants_1.DEFAULT_HOST_NAME,
        runMode: getRunMode_1.getRunMode(),
        jestLib,
        jestOpts: {
            projects,
            runInBand: runInBandJest,
            ...jestOpts,
        },
        logLevel,
        perfStart: Date.now(),
        runInBand,
        skipCheckConnection: false,
    };
};
